### PR TITLE
Formatter factory method

### PR DESCRIPTION
**Pull request by tienvx**
Depend on:
* https://github.com/pact-foundation/pact-php/pull/597
* https://github.com/pact-foundation/pact-php/pull/598

Apply `Bridge` and `Factory Method` design patterns to refactor formatters

Here is the full diagram:

matchers-formatters

### CODE REVIEW

File: example/csv/consumer/tests/Service/HttpClientServiceTest.php [hunk 1]
```diff
@@ -18,7 +17,7 @@ class HttpClientServiceTest extends TestCase
 {
     public function testGetCsvFile(): void
     {
-        $matcher = new Matcher(new PluginFormatter());
```
Comment (YOU54F) [inline example/csv/consumer/tests/Service/HttpClientServiceTest.php:21]: This looks like a nicer experience for end users (removing an additional import)

**Comment by YOU54F** [comment]
Nice change! It's a biggie so appreciate you splitting out the two PR's prior, as it helped me focus on them.

I'm looking at the end user interface changes in the examples, and that looks make it easier (removing an import and having as a option), considering your touched ~140 files, have minimal changes in your tests and everything green makes me a very confident person.

Appreciate the diagram!

So in a nutshell does this make it easier to maintainers/contributors to maintain/extend these interfaces in future? Whats the big selling point of moving to the new pattern.

Looks good to me though, haven't ran it locally but he have 💚 ci

File: example/csv/consumer/tests/Service/HttpClientServiceTest.php (hunk 1 above)
Comment (tienvx) [reply example/csv/consumer/tests/Service/HttpClientServiceTest.php:21]: :hugs: 

There is a small problem with these code: user need to figure out if a plugin is being used or not.

**Comment by tienvx** [comment]
> So in a nutshell does this make it easier to maintainers/contributors to maintain/extend these interfaces in future?

Yes, you are correct. This is the whole point of this PR.

**Comment by tienvx** [comment]
Self-reviewed
